### PR TITLE
#167940543 Implement Global Search

### PR DESCRIPTION
**Pull request by chialuka**
#### What does this PR do?
- This PR extends the search function to enable searching all the models

#### Description of Task proposed in this pull request?
- A global search function was added and it was configured to return results from the queries done on the database models

#### How should this be manually tested (Quality Assurance)?
- Make a `GET` request to `/api/v1/search/?global=query`. Please, change the query to the value you want to search for. You will get an aggregation of results from the different fields queried

#### What are the relevant pivotal tracker stories?
- [[167940543]](https://www.pivotaltracker.com/story/show/167940543)

#### Any background context you want to add (Operations Impact)?
- The original story for search did not make provision for a single search that will cut across all fields. This PR makes that provision

#### What I have learned working on this feature: 
- I learned how to implement a global search

#### Screenshots: 
Screenshot 2019-08-17 at 11 53 08 AM

### CODE REVIEW

File: server/database/models/article.js [hunk 1]
```diff
@@ -82,6 +83,119 @@ export default (sequelize, DataTypes) => {
...
+          comment: item.comment,
+          author: {
+            firstName: item.firstName,
+            lastName: item.lastName,
+            userName: item.userName,
+            avatarUrl: item.avatarUrl
+          }
+        });
+      }
+      delete item.comment;
```
Comment (dinobi) [inline server/database/models/article.js]: can we make lines 128 - 131 more idiomatic?, something like: `['comment', 'firstname', 'lastname', 'username'].forEach(prop => delete item[prop])`

File: test/search/search.test.js [hunk 2]
```diff
@@ -101,7 +104,37 @@ describe('Search Tests', () => {
...
+        .request(app)
+        .get(`${BASE_URL}/search/?article=dummy`);
+      expect(response).to.have.status(200);
+      expect(response.body).to.be.an('object');
+      expect(response.body.data.count).to.not.equal('0');
+    });
+  });
+
+  context('when the user enters an unkwown global search query', () => {
+    it('returns a body', async () => {
```
Comment (dinobi) [inline test/search/search.test.js]: I don't understand this test case, is the body supposed to contain value or not, let the test tell us about that

File: test/search/search.test.js [hunk 3]
```diff
@@ -101,7 +104,37 @@ describe('Search Tests', () => {
...
+        .request(app)
+        .get(`${BASE_URL}/search/?global=j`);
+      expect(response).to.have.status(200);
+      expect(response.body).to.be.an('object');
+      expect(response.body.userSearch.count).to.not.equal('0');
+    });
+  });
+
+  context('when the user enters a article search query', () => {
+    it('returns the search results', async () => {
```
Comment (dinobi) [inline test/search/search.test.js:179]: Can this test tell us more about the structure of the search result.

File: test/search/search.test.js [hunk 4]
```diff
@@ -101,7 +104,37 @@ describe('Search Tests', () => {
       expect(response).to.have.status(200);
       expect(response.body).to.be.an('object');
       expect(response.body.data.count).to.equal(0);
-      expect(response.body.itemsOnPage).to.equal(0);
+    });
+  });
+
+  context('when the user enters a global search query', () => {
+    it('returns the search results', async () => {
```
Comment (dinobi) [inline test/search/search.test.js:152]: pls tell us about the structure of the result

File: test/search/search.test.js [hunk 5]
```diff
@@ -35,20 +38,20 @@ describe('Search Tests', () => {
...
+      expect(response.body.data.results[0]).to.have.property('avatarUrl');
+      expect(response.body.data.results[0]).to.not.have.property('password');
+      expect(response.body.data.results[0]).to.not.have.property('role');
+      expect(response.body.data.results[0]).to.not.have.property('level');
     });
   });
 
   context('when the user enters a query for a nonexisting user', () => {
-    it('returns a 404 error', async () => {
+    it('returns a 200 error', async () => {
```
Comment (dinobi) [inline test/search/search.test.js]: A `200` error?

File: test/search/search.test.js (hunk 3 above)
Comment (chialuka) [reply test/search/search.test.js:179]: done

File: test/search/search.test.js (hunk 4 above)
Comment (chialuka) [reply test/search/search.test.js:152]: done

File: server/database/models/article.js (hunk 1 above)
Comment (chialuka) [reply server/database/models/article.js]: done

File: test/search/search.test.js (hunk 2 above)
Comment (chialuka) [reply test/search/search.test.js]: done.

**Review by dinobi — APPROVED** [review]

**Pull request merged**